secboot/secboot_tpm.go

```diff
@@ -384,17 +416,22 @@ func unlockEncryptedPartitionWithSealedKey(tpm *sb.TPMConnection, name, device,
 
 	// XXX: pinfile is currently not used
 	activated, err := sbActivateVolumeWithTPMSealedKey(tpm, name, device, keyfile, nil, &options)
-	if !activated {
-		// ActivateVolumeWithTPMSealedKey should always return an error if activated == false
-		return fmt.Errorf("cannot activate encrypted device %q: %v", device, err)
-	}
-	if err != nil {
-		logger.Noticef("successfully activated encrypted device %q using a fallback activation method", device)
-	} else {
-		logger.Noticef("successfully activated encrypted device %q with TPM", device)
-	}
 
-	return nil
+	if activated {
+		// non nil error may indicate the volume was unlocked using the
+		// recovery key
+		if err == nil {
+			logger.Noticef("successfully activated encrypted device %q with TPM", device)
+			return UnlockedWithSealedKey, nil
+		} else if isActivatedWithRecoveryKey(err) {
+			logger.Noticef("successfully activated encrypted device %q using a fallback activation method", device)
+			return UnlockedWithRecoveryKey, nil
+		}
+		// no other error is possible when activation succeeded
+		return NotUnlocked, fmt.Errorf("internal error: volume activated with unexpected error: %v", err)
```

Comment:
a bit unclearwhether we need a value Unknown for method in this case?

Reply:
If the current spread run is successful enough that we can merge a branch I'll push this in a followup.

Reply:
Pushed this now.